var/spack/repos/builtin/packages/magma/package.py

```diff
@@ -186,21 +186,40 @@ def post_install(self):
     def cache_test_sources(self):
         """Copy the example source files after the package is installed to an
         install test subdirectory for use during `spack test run`."""
-        self.cache_extra_test_sources([self.test_src_dir])
+        cache_extra_test_sources(self, [self.test_src_dir])
 
-    def test(self):
+    def test_c(self):
+        """Run benchmark tests"""
         test_dir = join_path(self.test_suite.current_test_cache_dir, self.test_src_dir)
-        with working_dir(test_dir, create=False):
-            pkg_config_path = "{0}/lib/pkgconfig".format(self.prefix)
+        with working_dir(test_dir):
+            pkg_config_path = self.prefix.lib.pkgconfig
             with spack.util.environment.set_env(PKG_CONFIG_PATH=pkg_config_path):
+
                 make("c")
-                self.run_test("./example_sparse", purpose="MAGMA smoke test - sparse solver")
-                self.run_test(
-                    "./example_sparse_operator", purpose="MAGMA smoke test - sparse operator"
-                )
-                self.run_test("./example_v1", purpose="MAGMA smoke test - legacy v1 interface")
-                self.run_test("./example_v2", purpose="MAGMA smoke test - v2 interface")
-                if "+fortran" in self.spec:
-                    make("fortran")
-                    self.run_test("./example_f", purpose="MAGMA smoke test - Fortran interface")
+                tests = [
+                    ("example_sparse", "sparse solver"),
+                    ("example_sparse_operator", "sparse operator"),
+                    ("example_v1", "legacy v1 interface"),
+                    ("example_v2", "v2 interface"),
+                ]
+
+                for test, desc in tests:
+                    with test_part(self, f"test_c_{test}", purpose=f"MAGMA smoke test - {desc}"):
+                        exe = which(test)
+                        exe()
+
+                make("clean")
+
+    def test_fortran(self):
+        """MAGMA smoke test - Fortran interface"""
```

Comment:
While this is the original description, it could be more concise.

```suggestion
        """Run Fortran example"""
```